src/core/compiler.jl

```diff
@@ -1,150 +1,126 @@
-using Base.Meta: parse
-
-# TODO: Replace this macro, see issue #514
 """
-Usage: @VarName x[1,2][1+5][45][3]
-  return: (:x,[1,2],6,45,3)
+Usage: @varname x[1,2][1+5][45][3]
+  return: VarName{:x}("[1,2][6][45][3]")
 """
-macro VarName(expr::Union{Expr, Symbol})
+macro varname(expr::Union{Expr, Symbol})
+    expr |> varname |> esc
+end
+function varname(expr)
     ex = deepcopy(expr)
-    isa(ex, Symbol) && return var_tuple(ex)
+    (ex isa Symbol) && return quote
+        Turing.VarName{$(QuoteNode(ex))}("")
+    end
     (ex.head == :ref) || throw("VarName: Mis-formed variable name $(expr)!")
     inds = :(())
     while ex.head == :ref
         if length(ex.args) >= 2
             strs = map(x -> :(string($x)), ex.args[2:end])
-            pushfirst!(inds.args, :("[" * join($(Expr(:vect, strs...)), ", ") * "]"))
+            pushfirst!(inds.args, :("[" * join($(Expr(:vect, strs...)), ",") * "]"))
         end
         ex = ex.args[1]
-        isa(ex, Symbol) && return var_tuple(ex, inds)
+        isa(ex, Symbol) && return quote
+            Turing.VarName{$(QuoteNode(ex))}(foldl(*, $inds, init = ""))
+        end
     end
     throw("VarName: Mis-formed variable name $(expr)!")
 end
-function var_tuple(sym::Symbol, inds::Expr=:(()))
-    return esc(:($(QuoteNode(sym)), $inds, $(QuoteNode(gensym()))))
-end
 
-
-function wrong_dist_errormsg(l)
-    return "Right-hand side of a ~ must be subtype of Distribution or a vector of " *
-        "Distributions on line $(l)."
+macro vsym(expr::Union{Expr, Symbol})
+    expr |> vsym
 end
-
-"""
-    generate_observe(observation, dist, model_info)
-
-Generate an observe expression for observation `observation` drawn from
-a distribution or a vector of distributions (`dist`).
-"""
-function generate_observe(observation, dist, model_info)
-    main_body_names = model_info[:main_body_names]
-    vi = main_body_names[:vi]
-    sampler = main_body_names[:sampler]
-    return quote
-        isdist = if isa($dist, AbstractVector)
-            # Check if the right-hand side is a vector of distributions.
-            all(d -> isa(d, Distribution), $dist)
-        else
-            # Check if the right-hand side is a distribution.
-            isa($dist, Distribution)
-        end
-        @assert isdist @error($(wrong_dist_errormsg(@__LINE__)))
-        $vi.logp += Turing.observe($sampler, $dist, $observation, $vi)
+function vsym(expr::Union{Expr, Symbol})
```

Comment:
Can you add an example of what this function does?

Reply:
```julia
expr = :(x[1,:][2])
vsym(expr)
```
returns `:x`.